plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,564 @@
+// Returns a the number of events that match the query.
+// Precondition: Query type is either `count` or `extract`.
+template <typename Actor>
+caf::expected<uint64_t>
+handle_lookup(Actor& self, const vast::query& query,
+              const std::shared_ptr<arrow::Table>& table) {
+  const auto& ids = query.ids;
+  const auto layout = type::from_arrow(*table->schema());
+  auto expr = expression{};
+  if (auto e = tailor(query.expr, layout); e) {
+    expr = *e;
+  }
+  uint64_t num_hits = 0ull;
+  auto record_batch_reader = arrow::TableBatchReader(*table);
+  auto handle_query = detail::overload{
+    [&](const query::count& count) {
+      if (count.mode == query::count::estimate)
+        die("logic error detected - count estimate should not load "
+            "partition");
+      std::shared_ptr<arrow::RecordBatch> batch{};
+      for (const auto& rb : record_batch_reader) {
+        if (rb.ok()) {
+          const auto slice = arrow_table_slice_builder::create(*rb);
+          auto result = count_matching(slice, expr, ids);
+          num_hits += result;
+          self->send(count.sink, result);
+        } else {
+          VAST_WARN("skipping record batch that failed to read: '{}'",
+                    rb.status());
+        }
+      }
+    },
+    [&](const query::extract& extract) {
+      if (extract.policy == query::extract::preserve_ids) {
+        return self->send(extract.sink, table_slice{});
+      }
+      for (const auto& rb : record_batch_reader) {
+        if (rb.ok()) {
+          const auto slice = arrow_table_slice_builder::create(*rb);
+          auto final_slice = filter(slice, expr, ids);
+          if (final_slice) {
+            num_hits += final_slice->rows();
+            self->send(extract.sink, *final_slice);
+          } else {
+            // TODO: error handling?
+            fmt::print(stderr, "tp;no slice today\n");
+          }
+        }
+      }
+    },
+  };
+  caf::visit(handle_query, query.cmd);
+  return num_hits;
+}
+
+std::shared_ptr<arrow::Schema> parse_arrow_schema_from_metadata(
+  const std::shared_ptr<parquet::FileMetaData>& parquet_metadata) {
+  if (!parquet_metadata)
+    return {};
+  auto arrow_metadata
+    = parquet_metadata->key_value_metadata()->Get("ARROW:schema");
+  if (!arrow_metadata.ok())
+    return {};
+  auto decoded = detail::base64::decode(*arrow_metadata);
+  auto schema_buf = std::make_shared<arrow::Buffer>(decoded);
+  arrow::ipc::DictionaryMemo dict_memo;
+  arrow::io::BufferReader input(schema_buf);
+  auto arrow_schema = arrow::ipc::ReadSchema(&input, &dict_memo);
+  if (!arrow_schema.ok())
+    return {};
+  return *arrow_schema;
+}
+
+caf::expected<std::shared_ptr<arrow::Table>>
+read_parquet_buffer(const chunk_ptr& chunk) {
+  const auto* ptr
+    = static_cast<const uint8_t*>(static_cast<const void*>(chunk->data()));
+  auto bufr = std::make_shared<arrow::io::BufferReader>(
+    ptr, detail::narrow_cast<int64_t>(chunk->size()));
```

Comment:
Is there an overload of the BufferReader constructor that takes an `arrow::Buffer`? If so, we should use that instead and convert the chunk into a buffer. If I recall correctly the buffer reader is then able to signal that it can be used without copying data. Not sure if Parquet makes use of that, but in any case the code would also be easier to read because you don't need all the pointer casting.

(Also note that the double cast in the line above should just be a single `reinterpret_cast`.)